File: 0207-course-schedule/0207-course-schedule.cpp
bool topo_sort(vector<vector<int>>& adj, int n) {
    vector<int> in(n,0);
    queue<int> q;
    for(int i=0; i<n; i++) {
        for(int num : adj[i]) {
            in[num] += 1;
        }
    }
    for(int i=0; i<n; i++) {
        if(in[i] == 0) q.push(i);
    }
    vector<int> ans;
    while(!q.empty()) {
        int a = q.front();
        q.pop();
        ans.push_back(a);
        for(int num : adj[a]) {
            in[num] -= 1;
            if(in[num] == 0) q.push(num);
        }
    }
    return ans.size() == n;
}


class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        int n = numCourses;
        vector<vector<int>> adj(n);
         for(auto &pre : prerequisites) {
            int course = pre[0];
            int prereq = pre[1];
            adj[prereq].push_back(course);  
        }
        bool res = topo_sort(adj,n);
        return res;
    }
};

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int n, int[][] nums) {
        ArrayList<ArrayList<Integer>> list = new ArrayList<>();
        for (int i = 0; i < n; i++) list.add(new ArrayList<>());
        int in[] = new int[n];
        for(int arr[] : nums) {
            list.get(arr[1]).add(arr[0]);
            in[arr[0]]++;
        }
        Queue<Integer> q = new LinkedList<>();
        for(int i=0; i<n; i++) {
            if(in[i] == 0) {
                q.offer(i);
            }
        }
        int count = 0;
        while(!q.isEmpty()) {
            int x = q.poll();
            count++;
            for(int num : list.get(x)) {
                in[num]--;
                if(in[num] == 0) q.offer(num);
            }
        }
        return count == n;
    }
}